keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt

```diff
@@ -63,150 +66,185 @@ class SecurityGroupHandler(
-    cloudDriverService.getSecurityGroup(resource.spec, resource.serviceAccount)
-
-  override suspend fun create(
-    resource: Resource<SecurityGroup>,
-    resourceDiff: ResourceDiff<SecurityGroup>
-  ): List<Task> {
-    val description: String
-    val taskRef = resource.spec.let { spec ->
-      description = "Create security group ${spec.moniker.name} in ${spec.accountName}/${spec.region}"
-      orcaService
-        .orchestrate(
-          resource.serviceAccount,
-          OrchestrationRequest(
-            description,
-            spec.moniker.app,
-            description,
-            listOf(spec.toCreateJob()),
-            OrchestrationTrigger(resource.id.toString())
+  override suspend fun desired(resource: Resource<SecurityGroup>): Map<String, SecurityGroup> =
+    with(resource.spec) {
+      locations.regions.map { region ->
+        Pair(
+          region,
+          SecurityGroup(
+            moniker = Moniker(app = moniker.app, stack = moniker.stack, detail = moniker.detail),
+            locations = SecurityGroup.Locations(accountName = locations.accountName, regions = setOf(region)),
+            vpcName = overrides[region]?.vpcName ?: vpcName,
+            description = overrides[region]?.description ?: description,
+            inboundRules = overrides[region]?.inboundRules ?: inboundRules,
+            overrides = emptyMap()
           ))
+      }.toMap()
     }
-    log.info("Started task {} to create security group", taskRef.ref)
-    return listOf(Task(id = taskRef.taskId, name = description))
-  }
 
-  override suspend fun update(
+  override suspend fun current(resource: Resource<SecurityGroup>): Map<String, SecurityGroup> =
+    cloudDriverService.getSecurityGroup(resource.spec, resource.serviceAccount)
+
+  override suspend fun upsert(
     resource: Resource<SecurityGroup>,
-    resourceDiff: ResourceDiff<SecurityGroup>
-  ): List<Task> {
-    val description: String
-    val taskRef = resource.spec.let { spec ->
-      description = "Update security group ${spec.moniker.name} in ${spec.accountName}/${spec.region}"
-      orcaService
-        .orchestrate(
-          resource.serviceAccount,
-          OrchestrationRequest(
-            description,
-            spec.moniker.app,
-            description,
-            listOf(spec.toUpdateJob()),
-            OrchestrationTrigger(resource.id.toString())
-          ))
+    resourceDiff: ResourceDiff<Map<String, SecurityGroup>>
+  ): List<Task> =
+    coroutineScope {
+      resourceDiff
+        .toIndividualDiffs()
+        .filter { diff -> diff.hasChanges() }
+        .map { diff ->
+          val spec = diff.desired
+          val job: Job
+          val verb: String
+
+          when (diff.current) {
+            null -> {
+              job = spec.toCreateJob()
+              verb = "Create"
+            }
+            else -> {
+              job = spec.toUpdateJob()
+              verb = "Update"
+            }
+          }
```

Comment:
A diff may include a mix of create and updates (i.e. when a new region is added to an existing resource definition); this ensures a newly added region gets the create treatment, required to support self-referential ingress rules.